src/dag.rs

```diff
@@ -388,28 +394,119 @@ impl DirectedAcyclicGraph {
+        vertex_count: vertex_count.into(),
+    }
+}
+
+#[derive(Debug)]
+pub struct DirectedAcyclicGraphStrategy {
+    vertex_count: Range<Vertex>,
+}
+
+#[derive(Debug)]
+pub struct DirectedAcyclicGraphValueTree {
+    vertex_count: u16,
+    vertex_mask_tree: DeltaDebuggingBitmapValueTree,
+    edge_bitmap_tree: DeltaDebuggingBitmapValueTree,
+    start_simplifying_edges: bool,
+}
+
+impl Strategy for DirectedAcyclicGraphStrategy {
+    type Tree = DirectedAcyclicGraphValueTree;
+
+    type Value = DirectedAcyclicGraph;
+
+    fn new_tree(
+        &self,
+        runner: &mut proptest::test_runner::TestRunner,
+    ) -> proptest::strategy::NewTree<Self> {
+        // Copied out of self.vertex_count.assert_nonempty(), because that's private to proptest
+        if self.vertex_count.is_empty() {
+            panic!(
+                "Invalid use of empty size range. (hint: did you \
+                 accidentally write {}..{} where you meant {}..={} \
+                 somewhere?)",
+                self.vertex_count.start,
+                self.vertex_count.end,
+                self.vertex_count.start,
+                self.vertex_count.end
+            );
+        }
+        let vertex_count =
+            Uniform::new(self.vertex_count.start, self.vertex_count.end - 1).sample(runner.rng());
+        let bitmap_size = strictly_upper_triangular_matrix_capacity(vertex_count);
+        let iter = (0..bitmap_size as u32).filter(|_| runner.rng().gen_bool(0.5));
+        let bitmap =
+            RoaringBitmap::from_sorted_iter(iter).map_err(|e| Reason::from(e.to_string()))?;
+        let vertex_mask = RoaringBitmap::from_sorted_iter(0..vertex_count as u32).unwrap();
+
+        Ok(DirectedAcyclicGraphValueTree {
+            vertex_count,
+            vertex_mask_tree: DeltaDebuggingBitmapValueTree::new(vertex_mask),
+            edge_bitmap_tree: DeltaDebuggingBitmapValueTree::new(bitmap),
+            start_simplifying_edges: false,
         })
-        .boxed();
+    }
+}
+
+impl ValueTree for DirectedAcyclicGraphValueTree {
+    type Value = DirectedAcyclicGraph;
+
+    fn current(&self) -> Self::Value {
+        let edge_map = self.edge_bitmap_tree.current();
+        let vertex_mask = self.vertex_mask_tree.current();
+
+        let mut from_dst = 0 as Vertex;
+        let mut edges = Vec::with_capacity(strictly_upper_triangular_matrix_capacity(
+            self.vertex_count,
+        ) as usize);
+        for from_src in 0..self.vertex_count {
+            if vertex_mask.contains(from_src as u32) {
+                let mut to_dst = from_dst + 1;
+                for to_src in from_src + 1..self.vertex_count {
+                    if vertex_mask.contains(to_src as u32) {
+                        let n = self.vertex_count as u32;
+                        let i = from_src as u32;
+                        let j = to_src as u32;
+                        let edge_idx = i * n + j - (((i + 1) * (i + 1) - (i + 1)) / 2 + (i + 1));
```

Comment:
This made me realize that during the dograph -> roaring-graphs transition, `StrictlyUpperTriangularLogicalMatrix` name lost its meaning since we're no longer using the "upper triangular" packing scheme.  I should rename it to `SparseSquareLogicalMatrix` since roaring bitmaps take care of storing only the relevant bits (literally bits).  I'm going to rename it in some future crate release.

In any case, this formula `i * n + j - (((i + 1) * (i + 1) - (i + 1)) / 2 + (i + 1));` made me pause for a bit.  Can we put it into a separate fn with a descriptive name, e.g. `strictly_upper_triangular_index` or similar?  🙏

Reply:
LOL yeah sorry for keeping that in, I meant to clean that up a bit :sweat_smile: 

Yeah, losing the packing scheme confused be a bunch, I initially wrote the code for dograph, so there was some head-scratching on my end.